digdag-standards/src/main/java/io/digdag/standards/operator/aws/EmrOperatorFactory.java

```diff
@@ -0,0 +1,1753 @@
+            }
+            return result.build();
+        }
+
+        private Submitter existingClusterSubmitter(AmazonElasticMapReduce emr, String tag, StepCompiler stepCompiler, String clusterId, Filer filer)
+        {
+            return () -> {
+                List<String> stepIds = pollingRetryExecutor(state, "submission")
+                        .retryUnless(AmazonServiceException.class, Aws::isDeterministicException)
+                        .withRetryInterval(DurationInterval.of(Duration.ofSeconds(30), Duration.ofMinutes(5)))
+                        .runOnce(new TypeReference<List<String>>() {}, s -> {
+
+                            RemoteFile runner = prepareRunner(filer, tag);
+
+                            // Compile steps
+                            stepCompiler.compile(runner);
+
+                            // Stage files to S3
+                            filer.stageFiles();
+
+                            AddJobFlowStepsRequest request = new AddJobFlowStepsRequest()
+                                    .withJobFlowId(clusterId)
+                                    .withSteps(stepCompiler.stepConfigs());
+
+                            int steps = request.getSteps().size();
+                            logger.info("Submitting {} EMR step(s) to {}", steps, clusterId);
+                            AddJobFlowStepsResult result = emr.addJobFlowSteps(request);
+                            logSubmittedSteps(clusterId, steps, i -> request.getSteps().get(i).getName(), i -> result.getStepIds().get(i));
+                            return ImmutableList.copyOf(result.getStepIds());
+                        });
+
+                return SubmissionResult.ofExistingCluster(clusterId, stepIds);
+            };
+        }
+
+        private void logSubmittedSteps(String clusterId, int n, Function<Integer, String> names, Function<Integer, String> ids)
+        {
+            logger.info("Submitted {} EMR step(s) to {}", n, clusterId);
+            for (int i = 0; i < n; i++) {
+                logger.info("Step {}/{}: {}: {}", i + 1, n, names.apply(i), ids.apply(i));
+            }
+        }
+
+        private Submitter newClusterSubmitter(AmazonElasticMapReduce emr, String tag, StepCompiler stepCompiler, Config clusterConfig, Filer filer, ParameterCompiler parameterCompiler)
+        {
+
+            return () -> {
+                // Start cluster
+                NewCluster cluster = pollingRetryExecutor(state, "submission")
+                        .withRetryInterval(DurationInterval.of(Duration.ofSeconds(30), Duration.ofMinutes(5)))
+                        // TODO: EMR requests are not idempotent, thus retrying might produce duplicate cluster submissions.
+                        .retryUnless(AmazonServiceException.class, Aws::isDeterministicException)
+                        .runOnce(NewCluster.class, s -> submitNewClusterRequest(emr, tag, stepCompiler, clusterConfig, filer, parameterCompiler));
+
```

Comment:
The cluster ID is stored in the state by `pollingRetryExecutor(state, "submission")` `runOnce`: https://github.com/treasure-data/digdag/blob/74691948f7cb3cc6af25ca406a287a13bdf09aff/digdag-standards/src/main/java/io/digdag/standards/operator/aws/EmrOperatorFactory.java#L466

So if `newClusterSubmitter` is re-entered due to bootstrap taking a long time (which it does), the submission is not performed again.